File: mvp-java/ui/src/main/java/com/wasisto/githubuserfinder/android/ui/search/SearchPresenter.java
package com.wasisto.githubuserfinder.android.ui.search;

import com.wasisto.githubuserfinder.common.Logger;
import com.wasisto.githubuserfinder.domain.models.SearchHistoryItem;
import com.wasisto.githubuserfinder.domain.models.SearchUsersResult;
import com.wasisto.githubuserfinder.domain.usecases.GetSearchHistoryUseCase;
import com.wasisto.githubuserfinder.domain.usecases.SearchUsersUseCase;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.ExecutorService;

public class SearchPresenter {

    private SearchView view;

    private SearchUsersUseCase searchUsersUseCase;

    private GetSearchHistoryUseCase getSearchHistoryUseCase;

    private Logger logger;

    private ExecutorService ioExecutorService;

    private ExecutorService mainExecutorService;

    private List<SearchHistoryItem> descSortedHistory;

    public SearchPresenter(SearchView view, SearchUsersUseCase searchUsersUseCase, GetSearchHistoryUseCase getSearchHistoryUseCase,
                           Logger logger, ExecutorService ioExecutorService, ExecutorService mainExecutorService) {
        this.view = view;
        this.getSearchHistoryUseCase = getSearchHistoryUseCase;
        this.searchUsersUseCase = searchUsersUseCase;
        this.logger = logger;
        this.ioExecutorService = ioExecutorService;
        this.mainExecutorService = mainExecutorService;
    }

    public void load() {
        mainExecutorService.execute(() -> {
            view.showLoadingIndicator();

            ioExecutorService.execute(() -> {
                try {
                    List<SearchHistoryItem> history = getSearchHistoryUseCase.execute();

                    logger.debug(String.format("history: %s", history));

                    descSortedHistory = new ArrayList<>(history);
                    Collections.sort(this.descSortedHistory, (item1, item2) -> Integer.compare(item2.getId(), item1.getId()));

                    mainExecutorService.execute(() -> {
                        view.showSearchHistory(descSortedHistory);
                    });
                } catch (Exception e) {
                    logger.warn("An error occurred while getting the search user history", e);

                    mainExecutorService.execute(() -> {
                        view.showErrorToast();
                    });
                }

                mainExecutorService.execute(() -> {
                    view.hideLoadingIndicator();
                });
            });
        });
    }

    public void search(String username) {
        mainExecutorService.execute(() -> {
            view.hideSearchHistory();
            view.showLoadingIndicator();

            ioExecutorService.execute(() -> {
                try {
                    SearchUsersResult result = searchUsersUseCase.execute(username);

                    logger.debug(String.format("result: %s", result));

                    mainExecutorService.execute(() -> {
                        if (result.getTotalCount() == 0) {
                            view.showNoResultsText();
                        } else {
                            view.hideNoResultsText();
                            view.showSearchResult(result.getItems());
                        }
                    });
                } catch (Exception e) {
                    logger.warn("An error occurred while searching users", e);

                    mainExecutorService.execute(() -> {
                        view.showSearchHistory(descSortedHistory);
                        view.showErrorToast();
                    });
                }

                mainExecutorService.execute(() -> {
                    view.hideLoadingIndicator();
                });
            });
        });
    }

    public void onResultItemClick(SearchUsersResult.Item resultItem) {
        mainExecutorService.execute(() -> {
            view.openUserDetailsActivity(resultItem.getLogin());
        });
    }

    public void onSearchHistoryItemClick(SearchHistoryItem historyItem) {
        mainExecutorService.execute(() -> {
            String query = historyItem.getQuery();
            view.setQuery(query);
            search(query);
        });
    }
}

File: mvp-kotlin/ui/src/main/java/com/wasisto/githubuserfinder/android/ui/search/SearchPresenter.kt
package com.wasisto.githubuserfinder.android.ui.search

import com.wasisto.githubuserfinder.common.Logger
import com.wasisto.githubuserfinder.domain.models.SearchHistoryItem
import com.wasisto.githubuserfinder.domain.models.SearchUsersResult
import com.wasisto.githubuserfinder.domain.usecases.GetSearchHistoryUseCase
import com.wasisto.githubuserfinder.domain.usecases.SearchUsersUseCase
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

class SearchPresenter(
    private val view: SearchView,
    private val searchUsersUseCase: SearchUsersUseCase,
    private val getSearchHistoryUseCase: GetSearchHistoryUseCase,
    private val logger: Logger,
    private val ioDispatcher: CoroutineDispatcher,
    private val mainDispatcher: CoroutineDispatcher,
    private val coroutineScope: CoroutineScope
) {

    private lateinit var descSortedHistory: List<SearchHistoryItem>

    fun load() = coroutineScope.launch(mainDispatcher) {
        view.showLoadingIndicator()

        launch(ioDispatcher) {
            try {
                val history: List<SearchHistoryItem> = getSearchHistoryUseCase()

                logger.debug("history: $history")

                descSortedHistory = history.sortedByDescending { it.id }

                launch(mainDispatcher) {
                    view.showSearchHistory(descSortedHistory)
                }
            } catch (e: Exception) {
                logger.warn("An error occurred while getting the search user history", e)

                launch(mainDispatcher) {
                    view.showErrorToast()
                }
            }

            launch(mainDispatcher) {
                view.hideLoadingIndicator()
            }
        }
    }

    fun search(username: String) = coroutineScope.launch(mainDispatcher) {
        view.hideSearchHistory()
        view.showLoadingIndicator()

        launch(ioDispatcher) {
            try {
                val result: SearchUsersResult = searchUsersUseCase(username)

                logger.debug("result: $result")

                launch(mainDispatcher) {
                    if (result.totalCount == 0L) {
                        view.showNoResultsText()
                    } else {
                        view.hideNoResultsText()
                        view.showSearchResult(result.items)
                    }
                }
            } catch (e: Exception) {
                logger.warn("An error occurred while searching users", e)

                launch(mainDispatcher) {
                    view.showSearchHistory(descSortedHistory)
                    view.showErrorToast()
                }
            }

            launch(mainDispatcher) {
                view.hideLoadingIndicator()
            }
        }
    }

    fun onResultItemClick(resultItem: SearchUsersResult.Item) = coroutineScope.launch(mainDispatcher) {
        view.openUserDetailsActivity(resultItem.login)
    }

    fun onSearchHistoryItemClick(historyItem: SearchHistoryItem) = coroutineScope.launch(mainDispatcher) {
        val query = historyItem.query
        view.setQuery(query)
        search(query)
    }
}
